File: Darts/src/burlton/dartzee/code/screen/stats/player/StatisticsTabGolfOptimalScorecard.java
package burlton.dartzee.code.screen.stats.player;
import burlton.core.code.obj.HandyArrayList;
import burlton.core.code.obj.HashMapList;
import burlton.dartzee.code.object.Dart;
import burlton.dartzee.code.screen.game.DartsScorerGolf;
import burlton.dartzee.code.stats.GameWrapper;

import javax.swing.*;
import java.awt.*;
import java.util.ArrayList;
import java.util.HashMap;

/**
 * Cherry-picks your best performance ever for each hole and assembles it into an 'ideal' scorecard.
 * Just for fun!
 */
public class StatisticsTabGolfOptimalScorecard extends AbstractStatisticsTab
{
	public StatisticsTabGolfOptimalScorecard() 
	{
		setLayout(new GridLayout(0, 2, 0, 0));
		add(panelMine);
		panelMine.setLayout(new BorderLayout(0, 0));
		panelOther.setLayout(new BorderLayout(0, 0));
		
		panelMine.add(panelMyScorecard);
		panelMyScorecard.setLayout(new BorderLayout(0, 0));
		add(panelOther);
		panelOther.add(panelOtherScorecard);
		panelOtherScorecard.setLayout(new BorderLayout(0, 0));
		
		DartsScorerGolf scorer = new DartsScorerGolf();
		scorer.init(null, null);
		panelMyScorecard.add(scorer, BorderLayout.CENTER);
	}
	
	private final JPanel panelMine = new JPanel();
	private final JPanel panelMyScorecard = new JPanel();
	private final JPanel panelOther = new JPanel();
	private final JPanel panelOtherScorecard = new JPanel();

	@Override
	public void populateStats()
	{
		setOtherComponentVisibility(this, panelOther);
		
		populateStats(filteredGames, panelMyScorecard, null);
		if (includeOtherComparison())
		{
			populateStats(filteredGamesOther, panelOtherScorecard, Color.RED);
		}
	}
	private void populateStats(HandyArrayList<GameWrapper> filteredGames, JPanel panel, Color color)
	{
		HashMapList<Integer, Dart> hmHoleToBestDarts = new HashMapList<>();
		HashMap<Integer, Long> hmHoleToBestGameId = new HashMap<>();
		for (GameWrapper game : filteredGames)
		{
			game.populateOptimalScorecardMaps(hmHoleToBestDarts, hmHoleToBestGameId);
		}
		
		DartsScorerGolf scorer = new DartsScorerGolf();
		if (color != null)
		{
			scorer.setTableForeground(Color.RED);
		}
		
		scorer.setShowGameId(true);
		scorer.init(null, null);
		
		for (int i=1; i<=18; i++)
		{
			ArrayList<Dart> darts = hmHoleToBestDarts.getAsHandyArrayList(i);
			if (darts != null)
			{
				long gameId = hmHoleToBestGameId.get(i);
				scorer.addDarts(darts, gameId);
			}
		}
		
		panel.removeAll();
		panel.add(scorer, BorderLayout.CENTER);
		panel.revalidate();
		panel.repaint();
	}

}

File: Darts/src/main/kotlin/dartzee/screen/stats/player/golf/StatisticsTabGolfOptimalScorecard.kt
package dartzee.screen.stats.player.golf

import dartzee.core.obj.HashMapList
import dartzee.`object`.Dart
import dartzee.screen.game.scorer.DartsScorerGolf
import dartzee.screen.stats.player.AbstractStatisticsTab
import dartzee.stats.GameWrapper
import java.awt.BorderLayout
import java.awt.Color
import java.awt.GridLayout
import javax.swing.JPanel

/**
 * Cherry-picks your best performance ever for each hole and assembles it into an 'ideal' scorecard.
 * Just for fun!
 */
class StatisticsTabGolfOptimalScorecard : AbstractStatisticsTab()
{
    private val panelMine = JPanel()
    private val panelMyScorecard = JPanel()
    private val panelOther = JPanel()
    private val panelOtherScorecard = JPanel()

    init
    {
        layout = GridLayout(0, 2, 0, 0)
        add(panelMine)
        panelMine.layout = BorderLayout(0, 0)
        panelOther.layout = BorderLayout(0, 0)

        panelMine.add(panelMyScorecard)
        panelMyScorecard.layout = BorderLayout(0, 0)
        add(panelOther)
        panelOther.add(panelOtherScorecard)
        panelOtherScorecard.layout = BorderLayout(0, 0)
        val scorer = DartsScorerGolf()
        scorer.init(null, null)
        panelMyScorecard.add(scorer, BorderLayout.CENTER)
    }

    override fun populateStats()
    {
        setOtherComponentVisibility(this, panelOther)

        populateStats(filteredGames, panelMyScorecard, null)
        if (includeOtherComparison())
        {
            populateStats(filteredGamesOther, panelOtherScorecard, Color.RED)
        }
    }

    private fun populateStats(filteredGames: List<GameWrapper>, panel: JPanel, color: Color?)
    {
        val hmHoleToBestDarts = HashMapList<Int, Dart>()
        val hmHoleToBestGameId = mutableMapOf<Int, Long>()

        val sortedGames = filteredGames.sortedBy { it.dtStart }
        for (game in sortedGames)
        {
            game.populateOptimalScorecardMaps(hmHoleToBestDarts, hmHoleToBestGameId)
        }

        val scorer = DartsScorerGolf()
        if (color != null)
        {
            scorer.setTableForeground(Color.RED)
        }

        scorer.showGameId = true
        scorer.init(null, null)

        for (i in 1..18)
        {
            val darts = hmHoleToBestDarts[i]
            if (darts != null)
            {
                val gameId = hmHoleToBestGameId[i]
                scorer.addDarts(darts, gameId!!)
            }
        }

        panel.removeAll()
        panel.add(scorer, BorderLayout.CENTER)
        panel.revalidate()
        panel.repaint()
    }

}
